### ISSUE TITLE
Handling file systems that cannot be remounted

### ISSUE THREAD

**Issue by nh2** [systemd/systemd#7007]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> 234

### Topic

Some file systems, for example GlusterFS, do not support `-o remount`.

However, when they are declared in `/etc/fstab` and systemd picsk them up, it still generates a `.mount` unit containing

```
ExecRemount=... -o remount,...
```

When that unit is reloaded, it'll fail.

Would it make sense to add an `x-systemd.*` option to the fstab entry that allows me to specify that remounts don't work (or shall not be used for some other administrative reason) for that file system, so that systemd instead executes a full unmount+mount instead?

Thanks!

**Comment by boucman** [comment]
naive question... the rootfs is either mounted by initrd (in which case it is not a remount, so you don't have the problem) or mounted by the kernel befor it starts PID1...

What is the point of having an entry for the rootfs in /etc/fstab if you can't remount it... what are you trying to achieve ?

**Comment by poettering** [comment]
> Would it make sense to add an x-systemd.* option to the fstab entry that allows me to specify that remounts don't work (or shall not be used for some other administrative reason) for that file system, so that systemd instead executes a full unmount+mount instead?

Adding such a mount optoin might make sense, but it should just remove the ExecReload= line, nothing more. Any automatism that converts reload → restart in this case can't really work, because restarts will impact everything depending on the unit (and for mount units that is likely a lot). Moreover we already have "systemctl reload-or-restart" for cases where people want this anyway.

**Comment by nh2** [comment]
> Adding such a mount optoin might make sense, but it should just remove the ExecReload= line, nothing more.

@poettering Do you mean `ExecRemount=` instead of `ExecReload=`?

What would `systemctl reload` do on such a unit, would it fail, or succeed but do nothing?

**Comment by poettering** [comment]
it would fail

**Comment by nh2** [comment]
@poettering I'm trying to look into that now, but somehow `reload-or-restart` doesn't seem to work as advertised for my mount unit.

From the man page:

```
       reload-or-restart PATTERN...
           Reload one or more units if they support it. If not, restart them instead. If
           the units are not running yet, they will be started.
```

I've switched from an fstab entry to a proper mount units with the following contents:

```
# systemctl cat mymount.mount
[Unit]
After=...
Requires=...
X-Restart-Triggers=...

[Mount]
TimeoutSec=5s
Type=glusterfs
What=10.0.0.1:/mymount
Where=/var/mymount/files
```

Since there's no `ExecRemount=` here (nor `ExecReload=`, I'm still not sure which one you meant above), I'd expect that systemd should `restart` this unit.

But when I run `systemctl reload-or-restart mymount.mount` I see in the journal:

```
Dec 17 00:12:40 node-1 systemd[1]: Reloading /var/mymount/files.
Dec 17 00:12:40 node-1 systemd[1]: mymount.mount: Mount process exited, code=exited status=1
Dec 17 00:12:40 node-1 systemd[1]: Reload failed for /var/mymount/files.
```

So it is still trying to `reload` instead of restart.

Is that expected?

Service units restart correctly if defined this way, it seems to be specific to mount units.

**Comment by arvidjaar** [comment]
> Is that expected?

Yes. `mount` units do support reload request and translate it internally into `mount -o remount`. It is just not documented.
